lib/ide/src/controller/module.rs

```diff
@@ -28,32 +56,161 @@ impl Location {
-    pub async fn file_externally_modified(&self) {
-        // TODO: notify underlying text/graph controllers about the changes
-        todo!()
+    /// Create a module controller for given location.
+    pub async fn new(location:Location, mut file_manager:fmc::Handle, mut parser:SharedParser)
+    -> FallibleResult<Self> {
+        let logger  = Logger::new(format!("Module Controller {}", location));
+        logger.info(|| "Loading module file");
+        let path    = location.to_path();
+        let content = file_manager.read(path).await?;
+        logger.info(|| "Parsing code");
+        let ast     = parser.parse(content,default())?;
+        logger.info(|| "Code parsed");
+        logger.trace(|| format!("The parsed ast is {:?}", ast));
+        let id_map  = default();
+        let data    = Controller {location,ast,file_manager,parser,id_map,logger};
+        Ok(Handle::new_from_data(data))
+    }
+
+    /// Save the module to file.
+    pub fn save_file(&self) -> impl Future<Output=Result<(),RpcError>> {
+        let (path,mut fm) = self.with_borrowed(|data| {
+            (data.location.to_path(),data.file_manager.clone_ref())
+        });
+        fm.write(path,self.code())
+        // TODO [ao] here save also the id_map and metadata.
+    }
+
+    #[cfg(test)]
+    fn new_mock
+    (location:Location, code:&str, id_map:IdMap, file_manager:fmc::Handle, mut parser:SharedParser)
+    -> FallibleResult<Self> {
+        let logger   = Logger::new("Mocked Module Controller");
+        let ast      = parser.parse(code.to_string(),id_map.clone())?;
+        let data     = Controller {location,ast,file_manager,parser,id_map,logger};
+        Ok(Handle::new_from_data(data))
+    }
+
+}
+
+
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use data::text::Span;
+    use data::text::Index;
+    use data::text::Size;
+    use json_rpc::test_util::transport::mock::MockTransport;
+    use uuid::Uuid;
+    use parser::SharedParser;
+    use basegl::display::shape::text::text_field::content::TextChange;
+    use basegl::display::shape::text::text_field::location::TextLocation;
+    use ast;
+    use wasm_bindgen_test::wasm_bindgen_test;
+    use ast::BlockLine;
+
+    #[wasm_bindgen_test]
+    fn update_ast_after_text_change() {
+        let transport    = MockTransport::new();
+        let file_manager = file_manager_client::Handle::new(transport);
+        let parser       = SharedParser::new().unwrap();
+        let location     = Location("Test".to_string());
+
+        let uuid1        = Uuid::new_v4();
+        let uuid2        = Uuid::new_v4();
+        let code         = "2+2";
+        let id_map       = IdMap(vec!
+            [ (Span::new(Index::new(0), Size::new(1)),uuid1.clone())
+            , (Span::new(Index::new(2), Size::new(1)),uuid2)
+            ]);
+
+        let controller   = Handle::new_mock(location,code,id_map,file_manager,parser).unwrap();
+
+        // Change number
```

Comment:
It's hard to grasp all these indices at a quick glance. Comment like "Change code to `22+2`" would help.